### ISSUE TITLE
Tree item bad indentation when adding items

### ISSUE THREAD

**Issue by matwachich** [fyne-io/fyne#3160]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

Let's say you have a Tree displaying some data. If you add data some items can lose their indentation : I mean a root item can become indented as if it was a child.

### How to reproduce

I made a code, just run it, and click the button to add item and see the result.

### Screenshots

<img width="451" alt="image" src="https://user-images.githubusercontent.com/1791230/180624316-0cee7338-a8be-4e03-b709-81da842cec89.png">

After clicking the Add Item button
<img width="451" alt="image" src="https://user-images.githubusercontent.com/1791230/180624334-791ded46-fc11-4dfe-97bd-70dd1c541cf8.png">

If I close the Group 1, and click Add Item one more time, the new item is correctly displayed
<img width="451" alt="image" src="https://user-images.githubusercontent.com/1791230/180624340-1a91c7fd-97b5-42c8-8ca9-c1c635cce237.png">

### Example code

```go
package main

import (
	"fmt"
	"strconv"
	"strings"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/widget"
)

type dataItem struct {
	Name  string
	Group string
}

func main() {
	a := app.New()
	w := a.NewWindow("Test")

	data := []dataItem{
		{Name: "Item 1", Group: ""},
		{Name: "Item 2", Group: ""},
		{Name: "Item 3", Group: ""},
		{Name: "Item 4", Group: "Group 1"},
		{Name: "Item 5", Group: "Group 1"},
		{Name: "Item 6", Group: "Group 1"},
		{Name: "Item 7", Group: "Group 2"},
	}

	tree := widget.NewTree(
		func(tni widget.TreeNodeID) (ret []widget.TreeNodeID) {
			if tni == "" {
				lastGroup := ""
				for i := 0; i < len(data); i++ {
					if lastGroup == "" && data[i].Group == lastGroup {
						ret = append(ret, fmt.Sprintf("item:%d", i))
						continue
					}
					if data[i].Group != lastGroup {
						ret = append(ret, fmt.Sprintf("group:%s", data[i].Group))
						lastGroup = data[i].Group
					}
				}
			} else {
				if len(tni) > 6 && tni[:6] == "group:" {
					for i := 0; i < len(data); i++ {
						if data[i].Group == tni[6:] {
							ret = append(ret, fmt.Sprintf("item:%d", i))
						}
					}
				}
			}
			return
		},
		func(tni widget.TreeNodeID) bool {
			return len(data) > 0 && (tni == "" || (len(tni) > 6 && tni[:6] == "group:"))
		},
		func(b bool) fyne.CanvasObject {
			return widget.NewLabel("...")
		},
		func(tni widget.TreeNodeID, b bool, co fyne.CanvasObject) {
			co.(*widget.Label).TextStyle.Bold = b
			if strings.HasPrefix(tni, "group:") {
				co.(*widget.Label).SetText(tni[6:])
			} else if strings.HasPrefix(tni, "item:") {
				idx, _ := strconv.Atoi(tni[5:])
				co.(*widget.Label).SetText(data[idx].Name)
			}
		},
	)

	w.SetContent(container.NewBorder(
		container.NewGridWithColumns(2,
			widget.NewButton("Add Item", func() {
				data = append(data[:3+1], data[3:]...)
				data[3].Name = "New Item"
				data[3].Group = ""
				tree.Refresh()
			}),
			widget.NewButton("Refresh", func() {
				tree.Refresh()
			}),
		),
		nil, nil, nil,
		tree,
	))

	w.Resize(fyne.NewSize(600, 400))
	w.CenterOnScreen()
	w.ShowAndRun()
}
```

### Fyne version

2.2.3

### Go compiler version

1.18.2

### Operating system

Windows

### Operating system version

10

### Additional Information

I feel like the problem is in this part of tree.go
![image](https://user-images.githubusercontent.com/1791230/180624402-57fb748b-a5b5-4177-a170-fc2ebd5c631d.png)

**Comment by ChandanChainani** [comment]
@matwachich The problem is due to uids are being generated again and again in below code
```
func(tni widget.TreeNodeID) (ret []widget.TreeNodeID) {
  if tni == "" {
    lastGroup := ""
    for i := 0; i < len(data); i++ {
      if lastGroup == "" && data[i].Group == lastGroup {
        ret = append(ret, fmt.Sprintf("item:%d", i))
        continue
      }
      if data[i].Group != lastGroup {
        ret = append(ret, fmt.Sprintf("group:%s", data[i].Group))
        lastGroup = data[i].Group
      }
    }
  } else {
    if len(tni) > 6 && tni[:6] == "group:" {
      for i := 0; i < len(data); i++ {
        if data[i].Group == tni[6:] {
          ret = append(ret, fmt.Sprintf("item:%d", i))
        }
      }
    }
  }
  return
}
```

In current implementation tree widget keeps track of nodes map to uid so that refresh/update/position only run on new nodes with new uid however the above code causes reassignment of existing uid to new node and because of which the new node does not get refreshed/updated/positioned

The problem will become more clear when we walk through code

when first `ChildUIDs` method is being called below uids will be returned
```
Root => [item:0 item:1 item:2 group:Group 1 group:Group 2]
Group1 => group:Group 1 [item:3 item:4 item:5]
Group2 => group:Group 2 [item:7]
```

In this case `item:3` is in `Group1` and it will be mapped to existing node in tree
when we add new node but clicking `Add Item` button again calls goes to `ChildUIDs` method
which again generate the uids which will look as below:
```
Root => [item:0 item:1 item:2 item:3 group:Group 1 group:Group 2]
Group1 => group:Group 1 [item:4 item:5 item:6]
Group2 => group:Group 2 [item:7]
```

As we can see the `item:3` uid got moved from `Group1` to `Root` and since it already exists no refresh/update/position happened just reassignment of new node to existing uid and after which `UpdateNode` method is called which update the text of all the nodes

To solve the problem you can maintain unique id in you `dataItem` type with ID field
```
type dataItem struct {
	ID    int
	Name  string
	Group string
}
```

And update the `ChildUIDs` and `UpdateNode` method accordingly. Below you can find code with suggested changes.
```go
data := []dataItem{
  {ID: 0, Name: "Item 1", Group: ""},
  {ID: 1, Name: "Item 2", Group: ""},
  {ID: 2, Name: "Item 3", Group: ""},
  {ID: 3, Name: "Item 4", Group: "Group 1"},
  {ID: 4, Name: "Item 5", Group: "Group 1"},
  {ID: 5, Name: "Item 6", Group: "Group 1"},
  {ID: 6, Name: "Item 7", Group: "Group 2"},
}

tree := widget.NewTree(
  func(tni widget.TreeNodeID) (ret []widget.TreeNodeID) {
    if tni == "" {
      lastGroup := ""
      for i := 0; i < len(data); i++ {
        if lastGroup == "" && data[i].Group == lastGroup {
          ret = append(ret, fmt.Sprintf("item:%d:%d", data[i].ID, i))
          continue
        }
        if data[i].Group != lastGroup {
          ret = append(ret, fmt.Sprintf("group:%s", data[i].Group))
          lastGroup = data[i].Group
        }
      }
    } else {
      if len(tni) > 6 && tni[:6] == "group:" {
        for i := 0; i < len(data); i++ {
          if data[i].Group == tni[6:] {
            ret = append(ret, fmt.Sprintf("item:%d:%d", data[i].ID, i))
          }
        }
      }
    }
    return
  },
  ...
  func(tni widget.TreeNodeID, b bool, co fyne.CanvasObject) {
    co.(*widget.Label).TextStyle.Bold = b
    if strings.HasPrefix(tni, "group:") {
      co.(*widget.Label).SetText(tni[6:])
    } else if strings.HasPrefix(tni, "item:") {
      arr := strings.Split(tni, ":")
      idx, _ := strconv.Atoi(arr[2])
      co.(*widget.Label).SetText(data[idx].Name)
    }
    co.(*widget.Label).Refresh()
  },
)
```

**Comment by andydotxyz** [comment]
Thanks so much @ChandanChainani this is a great answer.
Yes the uid of a tree node needs to be unique to the node (not just relative to the current parent) and is indeed used to persist a cache item. So an existing item in the tree should keep the same UID through refresh calls.

**Comment by matwachich** [comment]
Ok, I followed the advise, now I use a map, with string UUIDs that doesn't change (linked to data base primary key).

I keep having the same problem when an item changes depth: I'm storing items in categories. So an item can change categorie when edited.

**Comment by andydotxyz** [comment]
Can you provide a simple code sample we can see the problem?

**Comment by matwachich** [comment]
Sorry to reopen this, but after a while, where I totally forgot about this "bug", I encountered it again, and now since I understand better how Tree works, I've been able to make this simplification of how most of the Trees in my app are implemented.

You can see that I'm loading items from DB, each item has a UUID, a Title and a Category.
I load all of them, ordered by category and titre, in a map, and an ordering slice.

Items can have empty category, so they are displayed at the same level than the category parent item.
The problem arises when a no-category item becomes a categorized one (because those DB objects are editable). Item's UUID don't change, so for the Tree, it stills the same item. And the indentation is not updated.

First solution that I can think off (late night...) is to refactor all my Trees, and change namings of items in the data map and order slice (costy operation...).

Any thoughts please?

```go
package main

import (
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/widget"
	"github.com/google/uuid"
)

type dataItem struct {
	UID      uuid.UUID
	Category string
	Title    string
}

func main() {
	a := app.New()
	w := a.NewWindow("Tree Bug")

	data := make(map[uuid.UUID]dataItem)
	var order []uuid.UUID
	fnAddItem := func(category, title string) uuid.UUID {
		item := dataItem{
			UID:      uuid.New(),
			Category: category,
			Title:    title,
		}
		data[item.UID] = item
		order = append(order, item.UID)
		return item.UID
	}

	// empty categories must come first
	fnAddItem("", "Item 01")
	fnAddItem("", "Item 02")
	item03 := fnAddItem("Category 01", "Item 03")
	fnAddItem("Category 01", "Item 04")
	fnAddItem("Category 02", "Item 05")
	fnAddItem("Category 02", "Item 06")

	tree := widget.NewTree(
		func(tni widget.TreeNodeID) (ret []widget.TreeNodeID) {
			if tni == "" {
				lastCategory := ""
				for i := 0; i < len(order); i++ {
					if data[order[i]].Category == "" {
						ret = append(ret, data[order[i]].UID.String())
					} else {
						if data[order[i]].Category != lastCategory {
							ret = append(ret, "cat:"+data[order[i]].Category)
							lastCategory = data[order[i]].Category
						}
					}
				}
			} else {
				for i := 0; i < len(order); i++ {
					if data[order[i]].Category == tni[4:] {
						ret = append(ret, data[order[i]].UID.String())
					}
				}
			}
			return
		},
		func(tni widget.TreeNodeID) bool {
			return tni == "" || tni[:4] == "cat:"
		},
		func(b bool) fyne.CanvasObject { return widget.NewLabel("") },
		func(tni widget.TreeNodeID, b bool, co fyne.CanvasObject) {
			lbl := co.(*widget.Label)
			if tni[:4] == "cat:" {
				lbl.TextStyle.Bold = true
				lbl.Text = tni[4:]
			} else {
				lbl.TextStyle.Bold = false
				lbl.Text = data[uuid.MustParse(tni)].Title
			}
			co.Refresh()
		},
	)

	w.SetContent(container.NewBorder(
		widget.NewButton("Change Item 03 to no category", func() {
			item := data[item03]
			item.Category = ""
			data[item03] = item
			tree.Refresh()
		}),
		nil, nil, nil,
		tree,
	))
	w.ShowAndRun()
}
```

**Comment by andydotxyz** [comment]
> The problem arises when a no-category item becomes a categorized one (because those DB objects are editable). Item's UUID don't change, so for the Tree, it stills the same item. And the indentation is not updated.

That does sound like a possible issue - but I'm not sure it's the same as what was reported before?

**Comment by andydotxyz** [comment]
> Updating item's indentation at every update... will that be expensive op?

We will have to find a way to make this possible and performant...

**Comment by matwachich** [comment]
For now, I think I addressed this issue because I refactored all my DB logic (got rid of UUID, and now I have simple Autoincrementing INT RowIDs), and my TreeNodeIDs are of form [catID]:[itemID] ... so when an item changes its level, it changes its TreeNodeID.

I thought about a method tree.Reset in addition to .Refresh
Reset will do a total cache flush, then a Refresh.

This method could be added to other widgets too. Like tables, and Entries in order to force reset validation state (cf. some issue discussed in discord I think, where setting entry.SetText("") causes validation red).
